### ISSUE TITLE
CI: (ci-ipsec-upgrade) NodePort connectivity issues observed

### ISSUE THREAD

**Issue by giorio94** [cilium/cilium#29805]
## CI failure

Observed in different scenarios:
* https://github.com/cilium/cilium/actions/runs/7130505649/job/19513898236 (downgrade from v1.14 to v1.13 tip)
* https://github.com/cilium/cilium/actions/runs/7178332556/job/19546315226 (upgrade from v1.11 to v1.12 tip)
* https://github.com/cilium/cilium/actions/runs/7178332556/job/19546315390 (upgrade from v1.12.16 to v1.12 tip)
  * Additional failure for the `no-interrupted-connections` step; 
* https://github.com/cilium/cilium/actions/runs/7178347270/job/19546361178 (upgrade from v1.12 to v1.13 tip)
* https://github.com/cilium/cilium/actions/runs/7178347270/job/19546361383 (upgrade from v1.13.9 to v1.13 tip)

```
[.] Action [no-policies-extra/pod-to-local-nodeport/curl-3: cilium-test/client2-88575dbb7-twkt5 (10.244.3.201) -> cilium-test/echo-same-node (echo-same-node.cilium-test:8080)]
  [.] Action [no-policies-extra/pod-to-local-nodeport/curl-3: cilium-test/client2-88575dbb7-twkt5 (10.244.3.201) -> cilium-test/echo-same-node (echo-same-node.cilium-test:8080)]
  [-] Scenario [no-policies-extra/pod-to-remote-nodeport]
  [.] Action [no-policies-extra/pod-to-remote-nodeport/curl-0: cilium-test/client-75bff5f5b9-cvpnj (10.244.3.40) -> cilium-test/echo-other-node (echo-other-node.cilium-test:8080)]
  [.] Action [no-policies-extra/pod-to-remote-nodeport/curl-0: cilium-test/client-75bff5f5b9-cvpnj (10.244.3.40) -> cilium-test/echo-other-node (echo-other-node.cilium-test:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[fc00:f853:ccd:e793::4]:31452" failed: command terminated with exit code 56
```

The error repeats occurs multiple times until the action eventually times out, and at first glance it seems to always occur for IPv6 traffic. Sysdump are not linked here as huge.

**Comment by brb** [comment]
Happens on v1.13 w/ IPsec + vxlan + EP-routes=off when running `no-policies-extra/pod-to-remote-nodeport` for IPv6 requests when the nodeport request hits first an intermediate node.

<details>
<summary>The full tcpdump:</summary>

## client node

```
11:08:23.508697 lxc234e725d4095 In  IP6 fd00:10:244:2::dd22.34005 > fc00:c111::3.30042: Flags [S], seq 2104976501, win 32825, options [mss 1313,sackOK,TS val 3972578829 ecr 0,nop,wscale 7], length 0
11:08:23.508777 eth0  Out IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [S], seq 2104976501, win 32825, options [mss 1313,sackOK,TS val 3972578829 ecr 0,nop,wscale 7], length 0
11:08:23.509003 eth0  In  IP6 fc00:c111::3.30042 > fc00:c111::4.34005: Flags [S.], seq 535673466, ack 2104976502, win 32525, options [mss 1313,sackOK,TS val 2344177669 ecr 3972578829,nop,wscale 7], length 0
11:08:23.509026 lxc234e725d4095 Out IP6 fc00:c111::3.30042 > fd00:10:244:2::dd22.34005: Flags [S.], seq 535673466, ack 2104976502, win 32525, options [mss 1313,sackOK,TS val 2344177669 ecr 3972578829,nop,wscale 7], length 0
11:08:23.509044 lxc234e725d4095 In  IP6 fd00:10:244:2::dd22.34005 > fc00:c111::3.30042: Flags [.], ack 1, win 257, options [nop,nop,TS val 3972578829 ecr 2344177669], length 0
11:08:23.509057 eth0  Out IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [.], ack 1, win 257, options [nop,nop,TS val 3972578829 ecr 2344177669], length 0
11:08:23.509200 lxc234e725d4095 In  IP6 fd00:10:244:2::dd22.34005 > fc00:c111::3.30042: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972578830 ecr 2344177669], length 83
11:08:23.509217 eth0  Out IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972578830 ecr 2344177669], length 83
11:08:23.715327 lxc234e725d4095 In  IP6 fd00:10:244:2::dd22.34005 > fc00:c111::3.30042: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972579036 ecr 2344177669], length 83
11:08:23.715433 eth0  Out IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972579036 ecr 2344177669], length 83
11:08:23.715786 eth0  In  IP6 fc00:c111::3.30042 > fc00:c111::4.34005: Flags [R], seq 535673467, win 0, length 0
11:08:23.715827 lxc234e725d4095 Out IP6 fc00:c111::3.30042 > fd00:10:244:2::dd22.34005: Flags [R], seq 535673467, win 0, length 0
```

## intermediate node

```
11:08:23.508790 eth0  In  IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [S], seq 2104976501, win 32825, options [mss 1313,sackOK,TS val 3972578829 ecr 0,nop,wscale 7], length 0
11:08:23.508839 cilium_vxlan Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [S], seq 2104976501, win 32825, options [mss 1313,sackOK,TS val 3972578829 ecr 0,nop,wscale 7], length 0
11:08:23.508985 eth0  In  IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [S.], seq 535673466, ack 2104976502, win 32525, options [mss 1313,sackOK,TS val 2344177669 ecr 3972578829,nop,wscale 7], length 0
11:08:23.508999 eth0  Out IP6 fc00:c111::3.30042 > fc00:c111::4.34005: Flags [S.], seq 535673466, ack 2104976502, win 32525, options [mss 1313,sackOK,TS val 2344177669 ecr 3972578829,nop,wscale 7], length 0
11:08:23.509060 eth0  In  IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [.], ack 1, win 257, options [nop,nop,TS val 3972578829 ecr 2344177669], length 0
11:08:23.509071 cilium_vxlan Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [.], ack 1, win 257, options [nop,nop,TS val 3972578829 ecr 2344177669], length 0
11:08:23.509220 eth0  In  IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972578830 ecr 2344177669], length 83
11:08:23.509236 cilium_vxlan Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972578830 ecr 2344177669], length 83: HTTP: GET / HTTP/1.1
11:08:23.509306 eth0  In  IP6 fc00:c111::2.8080 > fc00:c111::3.34005: Flags [.], ack 2104976585, win 254, options [nop,nop,TS val 2344177670 ecr 3972578830], length 0
11:08:23.509326 eth0  Out IP6 fc00:c111::3.34005 > fc00:c111::2.8080: Flags [R], seq 2104976585, win 0, length 0
11:08:23.715461 eth0  In  IP6 fc00:c111::4.34005 > fc00:c111::3.30042: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972579036 ecr 2344177669], length 83
11:08:23.715517 cilium_vxlan Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972579036 ecr 2344177669], length 83: HTTP: GET / HTTP/1.1
11:08:23.715755 eth0  In  IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [R], seq 535673467, win 0, length 0
11:08:23.715780 eth0  Out IP6 fc00:c111::3.30042 > fc00:c111::4.34005: Flags [R], seq 535673467, win 0, length 0
```

## server node

```
11:08:23.508898 cilium_vxlan P   IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [S], seq 2104976501, win 32825, options [mss 1313,sackOK,TS val 3972578829 ecr 0,nop,wscale 7], length 0
11:08:23.508931 lxc6bb35f883dad Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [S], seq 2104976501, win 32825, options [mss 1313,sackOK,TS val 3972578829 ecr 0,nop,wscale 7], length 0
11:08:23.508954 lxc6bb35f883dad In  IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [S.], seq 535673466, ack 2104976502, win 32525, options [mss 1313,sackOK,TS val 2344177669 ecr 3972578829,nop,wscale 7], length 0
11:08:23.508982 eth0  Out IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [S.], seq 535673466, ack 2104976502, win 32525, options [mss 1313,sackOK,TS val 2344177669 ecr 3972578829,nop,wscale 7], length 0
11:08:23.509086 cilium_vxlan P   IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [.], ack 1, win 257, options [nop,nop,TS val 3972578829 ecr 2344177669], length 0
11:08:23.509098 lxc6bb35f883dad Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [.], ack 1, win 257, options [nop,nop,TS val 3972578829 ecr 2344177669], length 0
11:08:23.509250 cilium_vxlan P   IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972578830 ecr 2344177669], length 83: HTTP: GET / HTTP/1.1
11:08:23.509261 lxc6bb35f883dad Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972578830 ecr 2344177669], length 83: HTTP: GET / HTTP/1.1
11:08:23.509274 lxc6bb35f883dad In  IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [.], ack 84, win 254, options [nop,nop,TS val 2344177670 ecr 3972578830], length 0
11:08:23.509303 eth0  Out IP6 fc00:c111::2.8080 > fc00:c111::3.34005: Flags [.], ack 2104976585, win 254, options [nop,nop,TS val 2344177670 ecr 3972578830], length 0
11:08:23.509329 eth0  In  IP6 fc00:c111::3.34005 > fc00:c111::2.8080: Flags [R], seq 2104976585, win 0, length 0
11:08:23.509348 lxc6bb35f883dad Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [R], seq 2104976585, win 0, length 0
11:08:23.715615 cilium_vxlan P   IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972579036 ecr 2344177669], length 83: HTTP: GET / HTTP/1.1
11:08:23.715672 lxc6bb35f883dad Out IP6 fc00:c111::3.34005 > fd00:10:244:1::7de6.8080: Flags [P.], seq 1:84, ack 1, win 257, options [nop,nop,TS val 3972579036 ecr 2344177669], length 83: HTTP: GET / HTTP/1.1
11:08:23.715704 lxc6bb35f883dad In  IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [R], seq 535673467, win 0, length 0
11:08:23.715749 eth0  Out IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [R], seq 535673467, win 0, length 0
```
</details>

What happened is that the reply (after TCP handshake) from the server pod on the server node:
```
11:08:23.509274 lxc6bb35f883dad In  IP6 fd00:10:244:1::7de6.8080 > fc00:c111::3.34005: Flags [.], ack 84, win 254, options [nop,nop,TS val 2344177670 ecr 3972578830], length 0
11:08:23.509303 eth0  Out IP6 fc00:c111::2.8080 > fc00:c111::3.34005: Flags [.], ack 2104976585, win 254, options [nop,nop,TS val 2344177670 ecr 3972578830], length 0

```
escaped the vxlan tunnel and I got masqueraded by the following rule:
```
[107:10554] -A CILIUM_POST_nat -s fd00:10:244:1::/64 ! -d fd00:10:244:1::/64 ! -o cilium_+ -m comment --comment "cilium masquerade non-cluster" -j MASQUERADE
```

This is exactly what https://github.com/cilium/cilium/issues/23461 is about. The issue got fixed for >= v1.14 by changing the cilium_host's IPv6 addr.

I am going to disable IPv6 for this job.

**Comment by brb** [comment]
The backports got merged, closing.
